exec/java-exec/src/main/java/org/apache/drill/exec/ops/FragmentContext.java

```diff
@@ -17,483 +17,216 @@
-    this(dbContext, fragment, null, null, funcRegistry);
-  }
+  void waitForSendComplete();
 
   /**
-   * Create a FragmentContext instance for root fragment.
-   *
-   * @param dbContext DrillbitContext.
-   * @param fragment Fragment implementation.
-   * @param queryContext QueryContext.
-   * @param connection UserClientConnection.
-   * @param funcRegistry FunctionImplementationRegistry.
-   * @throws ExecutionSetupException
+   * Returns the UDF registry.
+   * @return the UDF registry
    */
-  public FragmentContext(final DrillbitContext dbContext, final PlanFragment fragment, final QueryContext queryContext,
-      final UserClientConnection connection, final FunctionImplementationRegistry funcRegistry)
-    throws ExecutionSetupException {
-    super(funcRegistry);
-    this.context = dbContext;
-    this.queryContext = queryContext;
-    this.connection = connection;
-    this.accountingUserConnection = new AccountingUserConnection(connection, sendingAccountor, statusHandler);
-    this.fragment = fragment;
-    contextInformation = new ContextInformation(fragment.getCredentials(), fragment.getContext());
-
-    logger.debug("Getting initial memory allocation of {}", fragment.getMemInitial());
-    logger.debug("Fragment max allocation: {}", fragment.getMemMax());
-
-    final OptionList list;
-    if (!fragment.hasOptionsJson() || fragment.getOptionsJson().isEmpty()) {
-      list = new OptionList();
-    } else {
-      try {
-        list = dbContext.getLpPersistence().getMapper().readValue(fragment.getOptionsJson(), OptionList.class);
-      } catch (final Exception e) {
-        throw new ExecutionSetupException("Failure while reading plan options.", e);
-      }
-    }
-    fragmentOptions = new FragmentOptionManager(context.getOptionManager(), list);
-
-    executionControls = new ExecutionControls(fragmentOptions, dbContext.getEndpoint());
-
-    // Add the fragment context to the root allocator.
-    // The QueryManager will call the root allocator to recalculate all the memory limits for all the fragments
-    try {
-      allocator = context.getAllocator().newChildAllocator(
-          "frag:" + QueryIdHelper.getFragmentId(fragment.getHandle()),
-          fragment.getMemInitial(),
-          fragment.getMemMax());
-      Preconditions.checkNotNull(allocator, "Unable to acuqire allocator");
-    } catch (final OutOfMemoryException e) {
-      throw UserException.memoryError(e)
-        .addContext("Fragment", getHandle().getMajorFragmentId() + ":" + getHandle().getMinorFragmentId())
-        .build(logger);
-    } catch(final Throwable e) {
-      throw new ExecutionSetupException("Failure while getting memory allocator for fragment.", e);
-    }
-
-    stats = new FragmentStats(allocator, fragment.getAssignment());
-    bufferManager = new BufferManagerImpl(this.allocator);
-    constantValueHolderCache = Maps.newHashMap();
-  }
+  FunctionImplementationRegistry getFunctionRegistry();
```

Comment:
OK to add as long as the registry can be created without ZK and network dependencies. (UDF introduced such dependencies, but, I believe, they can be turned off.)

Reply:
This is required since it's used by operators like join, hashagg, filter, and flatten. I've replaced FunctionImplementationRegistry with the FunctionLookupContext interface now. So it has no dependence on ZK or networking.